File: db/seed.js
// ============================================================
// seed.js — Populate the database from 996.json
// Usage: node db/seed.js
// ============================================================

import { readFileSync, existsSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import Database from 'better-sqlite3';

const __dirname = dirname(fileURLToPath(import.meta.url));
const DB_PATH = join(__dirname, 'beyond_bingo.db');
const DATA_PATH = join(__dirname, '..', 'core_logic', '996.json');

// Open DB
const db = new Database(DB_PATH);
db.pragma('journal_mode = WAL');
db.pragma('foreign_keys = ON');

// Read schema
const schema = readFileSync(join(__dirname, 'schema.sql'), 'utf-8');
db.exec(schema);

// Read game data
const raw = JSON.parse(readFileSync(DATA_PATH, 'utf-8'));
const gd = raw.gameData;

// ---- 1. Seed requirements ----
const insertReq = db.prepare(`
  INSERT OR IGNORE INTO requirements (id, type_id, name, display_name, prefix, suffix, helper_text)
  VALUES (?, ?, ?, ?, ?, ?, ?)
`);

const seenReq = new Set();
for (const cell of gd.remit) {
  for (const r of cell) {
    if (seenReq.has(r.id)) continue;
    seenReq.add(r.id);
    insertReq.run(r.id, r.type, r.name, r.displayName, r.prefix || null, r.suffix || null, r.helperText || null);
  }
}
console.log(`Seeded ${seenReq.size} requirements`);

// ---- 2. Seed players ----
const insertPlayer = db.prepare(`
  INSERT OR IGNORE INTO players (id, given_name, family_name, position)
  VALUES (?, ?, ?, ?)
`);

for (const p of gd.players) {
  insertPlayer.run(p.id, p.g || '', p.f, p.p || null);
}
console.log(`Seeded ${gd.players.length} players`);

// ---- 3. Seed player_requirements ----
const insertPR = db.prepare(`
  INSERT OR IGNORE INTO player_requirements (player_id, requirement_id)
  VALUES (?, ?)
`);

let prCount = 0;
for (const p of gd.players) {
  if (!p.v) continue;
  for (const rid of p.v) {
    insertPR.run(p.id, rid);
    prCount++;
  }
}
console.log(`Seeded ${prCount} player-requirement links`);

// ---- 4. Seed cells ----
const insertCell = db.prepare(`INSERT OR IGNORE INTO cells (id, label) VALUES (?, ?)`);
const insertCR = db.prepare(`INSERT OR IGNORE INTO cell_requirements (cell_id, requirement_id) VALUES (?, ?)`);

for (let i = 0; i < gd.remit.length; i++) {
  const cell = gd.remit[i];
  const label = cell.map(r => r.displayName).join(' + ');
  insertCell.run(i + 1, label);
  for (const r of cell) {
    insertCR.run(i + 1, r.id);
  }
}
console.log(`Seeded ${gd.remit.length} cells`);

// ---- 5. Seed the existing game (996) ----
const insertGame = db.prepare(`
  INSERT OR IGNORE INTO daily_games (game_id, date, seed) VALUES (?, ?, ?)
`);
const insertGC = db.prepare(`
  INSERT OR IGNORE INTO daily_game_cells (game_id, position, cell_id) VALUES (?, ?, ?)
`);
const insertGP = db.prepare(`
  INSERT OR IGNORE INTO daily_game_players (game_id, player_id) VALUES (?, ?)
`);

const tx = db.transaction(() => {
  insertGame.run(996, '2026-06-04', 42);

  for (let i = 0; i < gd.remit.length; i++) {
    insertGC.run(996, i, i + 1);
  }

  for (const p of gd.players) {
    insertGP.run(996, p.id);
  }
});
tx();
console.log('Seeded game 996');

db.close();
console.log('Done!');

File: db/seed.py
#!/usr/bin/env python3
"""
seed.py — Populate the SQLite database from 996.json
Zero dependencies. Uses Python's built-in sqlite3 module.

Usage: python3 db/seed.py
"""
import sqlite3
import json
import os

DB_PATH = os.path.join(os.path.dirname(__file__), 'beyond_bingo.db')
DATA_PATH = os.path.join(os.path.dirname(__file__), '..', 'core_logic', '996.json')
SCHEMA_PATH = os.path.join(os.path.dirname(__file__), 'schema.sql')

db = sqlite3.connect(DB_PATH)
db.execute("PRAGMA journal_mode=WAL")
db.execute("PRAGMA foreign_keys=ON")

# Read and execute schema
with open(SCHEMA_PATH) as f:
    db.executescript(f.read())

# Read game data
with open(DATA_PATH) as f:
    raw = json.load(f)
gd = raw['gameData']

# 1. Seed requirements
seen = set()
for cell in gd['remit']:
    for r in cell:
        if r['id'] in seen:
            continue
        seen.add(r['id'])
        db.execute(
            "INSERT OR IGNORE INTO requirements (id, type_id, name, display_name, prefix, suffix, helper_text) "
            "VALUES (?, ?, ?, ?, ?, ?, ?)",
            (r['id'], r['type'], r['name'], r['displayName'],
             r.get('prefix'), r.get('suffix'), r.get('helperText'))
        )
print(f"Seeded {len(seen)} requirements")

# 2. Seed players
for p in gd['players']:
    db.execute(
        "INSERT OR IGNORE INTO players (id, given_name, family_name, position) VALUES (?, ?, ?, ?)",
        (p['id'], p.get('g', ''), p['f'], p.get('p'))
    )
print(f"Seeded {len(gd['players'])} players")

# 3. Seed player_requirements
pr_count = 0
for p in gd['players']:
    for rid in p.get('v', []):
        db.execute(
            "INSERT OR IGNORE INTO player_requirements (player_id, requirement_id) VALUES (?, ?)",
            (p['id'], rid)
        )
        pr_count += 1
print(f"Seeded {pr_count} player-requirement links")

# 4. Seed cells
for i, cell in enumerate(gd['remit'], start=1):
    label = ' + '.join(r['displayName'] for r in cell)
    db.execute("INSERT OR IGNORE INTO cells (id, label) VALUES (?, ?)", (i, label))
    for r in cell:
        db.execute(
            "INSERT OR IGNORE INTO cell_requirements (cell_id, requirement_id) VALUES (?, ?)",
            (i, r['id'])
        )
print(f"Seeded {len(gd['remit'])} cells")

# 5. Seed game 996
db.execute("INSERT OR IGNORE INTO daily_games (game_id, date, seed) VALUES (?, ?, ?)",
           (996, '2026-06-04', 42))
for i in range(len(gd['remit'])):
    db.execute("INSERT OR IGNORE INTO daily_game_cells (game_id, position, cell_id) VALUES (?, ?, ?)",
               (996, i, i + 1))
for p in gd['players']:
    db.execute("INSERT OR IGNORE INTO daily_game_players (game_id, player_id) VALUES (?, ?)",
               (996, p['id']))
print("Seeded game 996")

db.commit()
db.close()
print("Done!")
